src/databricks/labs/ucx/progress/install.py

```diff
@@ -0,0 +1,75 @@
+import datetime as dt
+import logging
+from dataclasses import dataclass
+
+from databricks.labs.lsql.backends import Dataclass, SqlBackend
+from databricks.sdk.errors import InternalError
+from databricks.sdk.retries import retried
+
+from databricks.labs.ucx.framework.utils import escape_sql_identifier
+
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True, kw_only=True)
+class HistoricalRecord:
+    workspace_id: int
+    """The identifier of the workspace where this record was generated."""
+
+    run_id: int
+    """An identifier of the workflow run that generated this record."""
+
+    snapshot_id: int
+    """An identifier that is unique to the records produced for a given snapshot."""
+
+    run_start_time: dt.datetime
+    """When this record was generated."""
+
+    object_type: str
+    """The inventory table for which this record was generated."""
+
+    object_type_version: int
+    """Versioning of inventory table, for forward compatibility."""
+
+    object_id: list[str]
+    """The type-specific identifier for this inventory record."""
+
+    object_data: str
```

Comment:
```suggestion
    data: dict[str,str]
```
we'll impose a constraint on the writer, so that readers are easier.

Reply:
`klazz(**object_data)` works just fine.

Reply:
We haven't covered what the values in the map should contain.

Until recently our inventory classes were 'flat', and a common workaround is to have a field that encodes structured data:
 * Of the 19 inventory classes, 7 have a `failures` field which is a JSON-encoded array.[^1]
 * `Permissions` has a generic `raw` field that contains arbitrary JSON-encoded data.
 * `ReconResult` has `schema_comparison` and `data_comparison`: also JSON-encoded.
 * `SubmitRunInfo` has JSON-encoded `run_ids` (also mentioned elsewhere in the discussion object identifiers).
 * `Table` has `storage_properties`, which is a custom-encoded `dict[str,str]`.

The recent exception to this is `DirectFsAccess`, which is the first non-flat data inventory class using the recently introduced support in LSQL for complex data types.

With this in mind, there are a few options for what the values in the `object_data` map should be:
 * Blindly JSON-encode it. This will work, and is simple. The main caveats are that: string values will be quoted, and the JSON-encoded fields will end up double-encoded.
 * As above, but: if the field type is string, store as-is. This eliminates unnecessary encoding of all the string fields. The JSON-encoded fields remain as such.

@nfx: How does this sound? Does anything else come to mind?

[^1]: Elsewhere we propose removing this field from classes that have it and storing this directly on the `failures` attribute.

Reply:
@nfx: Please respond to the above

Reply:
1) `failures` has to become a top-level field, because we'll perform many different SQL operations on it (grouping, counting, distinct, etc) and it shouldn't get to `data: dict[str,str]` field
2) we don't sync `Permissions`, `ReconResult`, and `SubmitRunInfo` to multiworkspace database
3) to simplify things, let's not put `storage_properties` as we're not actually relying on it for anything.

to get to type safety, each dataclass instance that needs syncing has to get an explicit method

```
def to_historical(self) -> HistoricalRecord:
  return HistoricalRecord(...)
```
and some of the classes, like `Table`, would need a `from_historical` class method:

```
@classmethod
def from_historical(cls, h: HistoricalRecord):
  ...
```

Reply:
@nfx: That's clear.

We still need to define the semantics of `object_data` values to cover the `source_lineage` field of the `SourceInfo` subclasses: `DirectFsAccess` and `UsedTable`.

Options are:

 - JSON encoded, as suggested above.
 - Ignored (like `failures`)

Or maybe something else…?